src/Test/Internal.elm

```diff
@@ -99,28 +119,65 @@ fuzzTest fuzzer desc getExpectation =
             case Lazy.List.headAndTail branches of
-                Just ( (Rose branch _) as rosetree, moreLazyRoseTrees ) ->
+                Just ( (Rose possiblyFailingValue _) as rosetree, moreLazyRoseTrees ) ->
                     -- either way, recurse with the most recent failing expectation, and failing input with its list of shrunken values
-                    case getExpectation branch of
-                        Pass ->
-                            shrink oldExpectation (Rose root moreLazyRoseTrees)
-
-                        newExpectation ->
-                            shrink newExpectation rosetree
+                    case Dict.get (toString possiblyFailingValue) results of
+                        Just result ->
+                            case result of
+                                Passes ->
+                                    shrink oldExpectation (Rose failingValue moreLazyRoseTrees) results
+
+                                ShrinksTo minimalValue ->
+                                    ( minimalValue, oldExpectation, results )
+
+                        Nothing ->
+                            case getExpectation possiblyFailingValue of
+                                Pass ->
+                                    shrink oldExpectation
+                                        (Rose failingValue moreLazyRoseTrees)
+                                        (Dict.insert (toString possiblyFailingValue) Passes results)
+
+                                newExpectation ->
+                                    let
+                                        ( minimalValue, finalExpectation, results_ ) =
+                                            shrink newExpectation rosetree results
+                                    in
+                                        ( minimalValue
+                                        , finalExpectation
+                                        , results
+                                            |> Dict.insert (toString possiblyFailingValue) (ShrinksTo minimalValue)
+                                            |> Dict.insert (toString failingValue) (ShrinksTo minimalValue)
+                                        )
 
                 Nothing ->
-                    ( root, oldExpectation )
+                    ( failingValue
+                    , oldExpectation
+                    , Dict.insert (toString failingValue) (ShrinksTo failingValue) results
+                      -- minimal value!
+                    )
+
+        (Rose failingValue _) =
+            rootTree
 
-        ( result, finalExpectation ) =
-            shrink rootsExpectation rootTree
+        ( minimalValue, finalExpectation, results_ ) =
+            shrink rootsExpectation rootTree results
     in
-        Dict.insert (toString result) finalExpectation dict
+        ( Dict.insert (toString minimalValue) finalExpectation failures
+        , Dict.insert (toString failingValue) (ShrinksTo minimalValue) results
```

Comment:
Is it correct to use `results` and not `results_` here? Could the underscore be changed to `newResults` or something more meaningful?